helpers/getLocalizeResources.js

```diff
@@ -0,0 +1,355 @@
+import { getDocumentLocaleSettings } from '@brightspace-ui/intl/lib/common.js';
+
+const CacheName = 'd2l-oslo';
+const ContentTypeHeader = 'Content-Type';
+const ContentTypeJson = 'application/json';
+const DebounceTime = 150;
+const ETagHeader = 'ETag';
+const OverrideLanguage = 'overrides';
+const StateFetching = 2;
+const StateIdle = 1;
+
+const BatchFailedReason = new Error('Failed to fetch batch overrides.');
+const SingleFailedReason = new Error('Failed to fetch overrides.');
+
+const blobs = new Map();
+
+let cache = undefined;
+let cachePromise = undefined;
+let documentLocaleSettings = undefined;
+let queue = [];
+let state = StateIdle;
+let timer = 0;
+
+async function publish(request, response) {
+
+	if (response.ok) {
+		const blob = await response.blob();
+		const objectUrl = URL.createObjectURL(blob);
+		request.resolve(objectUrl);
+	} else {
+		request.reject(SingleFailedReason);
+	}
+}
+
+async function flushQueue() {
+
+	timer = 0;
+	state = StateFetching;
+
+	if (queue.length <= 0) {
+		state = StateIdle;
+		return;
+	}
+
+	const requests = queue;
+
+	queue = [];
+
+	const resources = requests.map(item => item.resource);
+	const bodyObject = { resources };
+	const bodyText = JSON.stringify(bodyObject);
+
+	const res = await fetch(documentLocaleSettings.oslo.batch, {
+		method: 'POST',
+		body: bodyText,
+		headers: { [ContentTypeHeader]: ContentTypeJson }
+	});
+
+	if (res.ok) {
+
+		const responses = (await res.json()).resources;
+
+		const tasks = [];
+
+		for (let i = 0; i < responses.length; ++i) {
+
+			const response = responses[i];
+			const request = requests[i];
+
+			const responseValue = new Response(response.body, {
+				status: response.status,
+				headers: response.headers
+			});
+
+			// New version might be available since the page loaded, so make a
+			// record of it.
+
+			const nextVersion = responseValue.headers.get(ETagHeader);
+			if (nextVersion) {
+				setVersion(nextVersion);
+			}
+
+			const cacheKey = new Request(formatCacheKey(request.resource));
+			const cacheValue = responseValue.clone();
+
+			if (cache === undefined) {
+				if (cachePromise === undefined) {
+					cachePromise = caches.open(CacheName);
+				}
+				cache = await cachePromise;
+			}
+
+			console.log(`[Oslo] cache prime: ${request.resource}`);
+			tasks.push(cache.put(cacheKey, cacheValue));
+			tasks.push(publish(request, responseValue));
+		}
+
+		await Promise.all(tasks);
+
+	} else {
+
+		for (const request of requests) {
+
+			request.reject(BatchFailedReason);
+		}
+	}
+
+	if (queue.length > 0) {
+		setTimeout(flushQueue, 0);
```

Comment:
In this case, more requests came in while we were processing the queue... so why don't we let the normal batching process handle them? i.e. why process them immediately? I guess it's because `debounceQueue` bails if the state isn't Idle...

Reply:
Yeah, I didn't want to add an artificial delay if queuing was already happening while the request was in flight. This might be tweakable after if the model component (FACE) doesn't correspond well to other usages.

Reply:
Cool makes sense, thx.